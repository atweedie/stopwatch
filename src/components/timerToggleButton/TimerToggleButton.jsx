import { useContext } from "react"
import { TimerState } from "../../modules/timerStateContext"

const TimerToggleButton = () => {
    const {isRunning, setIsRunning} = useContext(TimerState);

    return (
        <button className="button" id="stopwatch-start-stop" onClick={() => setIsRunning(!isRunning)}>
            {isRunning ? 'Stop' : 'Start'}
        </button>
    )
}

export default TimerToggleButton