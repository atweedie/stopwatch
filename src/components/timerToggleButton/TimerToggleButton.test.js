import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react';
import TimerToggleButton from './TimerToggleButton';

const stubSetElapsedTime = jest.fn();
const stubSetIsRunning = jest.fn();
const stubUseContext = jest.fn();

const defaultContext = {
    isRunning: false,
    setElapsedTime: stubSetElapsedTime,
    setIsRunning: stubSetIsRunning
}

const renderWithContext = (context = defaultContext) => {
    React.useContext = stubUseContext.mockReturnValue(context)
    return render(<TimerToggleButton/>)
};

beforeEach(() => {
    jest.useFakeTimers();
})

afterEach(() => {
    jest.useRealTimers();
});

test('should render as expected when isRunning is set to false', () => {
const component = renderWithContext();
    
    expect(component.container).toMatchSnapshot();
});

test('should render as expected when isRunning is set to true', () => {
    const givenContext = {
        ...defaultContext,
        isRunning: true
    }

    const component = renderWithContext(givenContext);
        
        expect(component.container).toMatchSnapshot();
    });

test('should call setIsRunning with true on button click when isRunning is false', () => {
    renderWithContext();

    const button = screen.getByText('Start');
        
    fireEvent.click(button)

    jest.runAllTimers();

    expect(stubSetIsRunning).toHaveBeenCalledWith(true);
});

test('should call setIsRunning with false on button click when isRunning is true', () => {
    const givenContext = {
        ...defaultContext,
        isRunning: true
    }
    
    renderWithContext(givenContext);

    const button = screen.getByText('Stop');
        
    fireEvent.click(button)

    jest.runAllTimers();

    expect(stubSetIsRunning).toHaveBeenCalledWith(false);
});