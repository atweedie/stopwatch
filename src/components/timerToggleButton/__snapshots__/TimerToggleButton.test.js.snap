// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as expected when isRunning is set to false 1`] = `
<div>
  <button
    class="button"
    id="stopwatch-start-stop"
  >
    Start
  </button>
</div>
`;

exports[`should render as expected when isRunning is set to true 1`] = `
<div>
  <button
    class="button"
    id="stopwatch-start-stop"
  >
    Stop
  </button>
</div>
`;
